Source/Core/Common/Config/ConfigInfo.h

```diff
@@ -42,12 +65,64 @@ struct Info
   template <typename Enum,
             std::enable_if_t<std::is_same<T, detail::UnderlyingType<Enum>>::value>* = nullptr>
   Info(const Info<Enum>& other)
-      : location{other.location}, default_value{static_cast<detail::UnderlyingType<Enum>>(
-                                      other.default_value)}
+      : m_location{other.GetLocation()}, m_default_value{static_cast<T>(other.GetDefaultValue())},
+        m_cached_value{other.template GetCachedValueCasted<T>()}
+  {
+  }
+
+  Info<T>& operator=(const Info<T>& other)
+  {
+    m_location = other.GetLocation();
+    m_default_value = other.GetDefaultValue();
+    m_cached_value = other.GetCachedValue();
+  }
+
+  Info<T>& operator=(Info<T>&& other)
+  {
+    m_location = std::move(other.m_location);
+    m_default_value = std::move(other.m_default_value);
+    m_cached_value = std::move(other.m_cached_value);
+  }
+
+  // Make it easy to convert Info<Enum> into Info<UnderlyingType<Enum>>
+  // so that enum settings can still easily work with code that doesn't care about the enum values.
+  template <typename Enum,
+            std::enable_if_t<std::is_same<T, detail::UnderlyingType<Enum>>::value>* = nullptr>
+  Info<T>& operator=(const Info<Enum>& other)
   {
+    m_location = other.GetLocation();
+    m_default_value = static_cast<T>(other.GetDefaultValue());
+    m_cached_value = other.template GetCachedValueCasted<T>();
   }
 
-  Location location;
-  T default_value;
+  constexpr const Location& GetLocation() const { return m_location; }
+  constexpr const T& GetDefaultValue() const { return m_default_value; }
+
+  CachedValue<T> GetCachedValue() const
+  {
+    std::unique_lock lock(m_cached_value_mutex);
+    return m_cached_value;
+  }
+
+  template <typename U>
+  CachedValue<U> GetCachedValueCasted() const
+  {
+    std::unique_lock lock(m_cached_value_mutex);
+    return CachedValue<U>{static_cast<U>(m_cached_value.value), m_cached_value.config_version};
+  }
+
+  void SetCachedValue(const CachedValue<T>& cached_value) const
+  {
+    std::unique_lock lock(m_cached_value_mutex);
+    if (m_cached_value.config_version < cached_value.config_version)
+      m_cached_value = cached_value;
+  }
+
+private:
+  Location m_location;
+  T m_default_value;
+
+  mutable CachedValue<T> m_cached_value;
+  mutable std::mutex m_cached_value_mutex;
```

Comment:
A good opportunity to use std::shared_mutex?

Reply:
I wasn't aware that that class existed. It does sound like a good fit. (Not that I think we access the same setting from multiple threads often at all.)